src/driver-roccat-kone-emp.c

```diff
@@ -0,0 +1,1127 @@
+ * The LED effects BLINKING and PULSING are not supported in libratbag
+ * The maximum macro size is 480 in the mouse software. One event keeps the event data and the timing/delay
+	- libratbag does not keep track of that number of events. It limits the mouse to 128 events
+ * The mouse can repeat macro. Not supported in libratbag
+ * In official soft, we can set a LED color to offset the cycle effect (only with predefined_led_colors). 
+ *   Since predefined colors are not handled, we can't reproduce this effect.
+ */
+#include "config.h"
+#include <assert.h>
+#include <bits/stdint-uintn.h>
+#include <errno.h>
+#include <libevdev/libevdev.h>
+#include <linux/input.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <string.h>
+#include <math.h>
+
+#include "libratbag-enums.h"
+#include "libratbag-private.h"
+#include "libratbag-hidraw.h"
+#include "libratbag.h"
+#include "shared-macro.h"
+
+#define ROCCAT_PROFILE_MAX          5
+#define ROCCAT_BUTTON_MAX           11 * 2 // (Easy Shift)
+#define ROCCAT_NUM_DPI              5
+#define ROCCAT_LED_MAX              4
+
+#define ROCCAT_MAX_RETRY_READY      10
+
+#define ROCCAT_REPORT_ID_CONFIGURE_PROFILE  4
+#define ROCCAT_REPORT_ID_PROFILE            5
+#define ROCCAT_REPORT_ID_SETTINGS           6
+#define ROCCAT_REPORT_ID_KEY_MAPPING        7
+#define ROCCAT_REPORT_ID_MACRO              8
+
+#define ROCCAT_MAGIC_NUMBER_SETTINGS        0x29
+#define ROCCAT_MAGIC_NUMBER_KEY_MAPPING     0x47
+
+#define ROCCAT_BANK_ID_1    1
+#define ROCCAT_BANK_ID_2    2
+#define ROCCAT_REPORT_SIZE_MACRO_BANK       1026
+
+#define ROCCAT_MACRO_GROUP_NAME_LENGTH  40
+#define ROCCAT_MACRO_NAME_LENGTH        32
+
+#define ROCCAT_CONFIG_SETTINGS      0x80 // (LED and mouse configuration)
+#define ROCCAT_CONFIG_KEY_MAPPING   0x90 // (Buttons configuration)
+
+#define ROCCAT_MAX_MACRO_LENGTH     480
+
+#define ROCCAT_MIN_DPI  100
+#define ROCCAT_MAX_DPI  12000
+
+#define ROCCAT_USER_DEFINED_COLOR   0x1e // The mouse knows some predefined colors. User can also set RGB values
+#define ROCCAT_LED_BLINKING         0x02
+#define ROCCAT_LED_BREATHING        0x03
+#define ROCCAT_LED_PULSING          0x04
+
+unsigned int report_rates[] = { 125, 250, 500, 1000 };
+
+struct color {
+	uint8_t r;
+	uint8_t g;
+	uint8_t b;
+} __attribute__((packed));
+struct color predefined_led_colors[] = { { 179, 0, 0 }, { 255, 0, 0 }, { 255, 71, 0}, { 255, 106, 0 },
+										 { 255, 157, 71 }, { 248, 232, 0 }, { 246, 255, 78 }, { 201, 255, 78 },
+										 { 185, 255, 78 }, { 132, 255, 78 }, { 0, 255, 0 }, { 0, 207, 55 },
+										 { 0, 166, 44 }, { 0, 207, 124 }, { 0,207, 158 }, { 0, 203, 207 },
+										 { 41, 197, 255 }, { 37, 162, 233 }, { 99, 158, 239 }, { 37, 132, 233 },
+										 { 0, 72, 255 }, { 15, 15, 255 }, { 15, 15, 188 }, { 89, 7, 255 },
+										 { 121, 12, 255 }, { 161, 12, 255 }, { 170, 108, 232 }, { 181, 10, 216 },
+										 { 205, 10, 217 }, { 217, 10, 125 } };
```

Comment:
looks like indentation is messed up here, might be github's diff view. Either way, better to make this easier with:
```c
struct color predefined_led_colors[] = {
  { 123, ... }
};
````